automerge-backend/src/protocol.rs

```diff
@@ -87,63 +87,58 @@ impl ActorID {
 pub struct Clock(pub HashMap<ActorID, u32>);
 
 impl Clock {
-    pub fn new(actor_id: &ActorID, seq: u32) -> Clock {
-        let mut map = HashMap::new();
-        map.insert(actor_id.clone(), seq);
-        Clock(map)
-    }
-
     pub fn empty() -> Clock {
         Clock(HashMap::new())
     }
 
-    pub fn with_dependency(&self, actor_id: &ActorID, new_seq: u32) -> Clock {
-        let mut result = self.0.clone();
-        result.insert(actor_id.clone(), new_seq);
-        Clock(result)
+    pub fn with(&self, actor_id: &ActorID, seq: u32) -> Clock {
+        let mut result = self.clone();
+        result.set(actor_id, max(seq, self.get(actor_id)));
+        result
     }
 
-    pub fn upper_bound(&self, other: &Clock) -> Clock {
-        let mut result: HashMap<ActorID, u32> = HashMap::new();
-        self.0.iter().for_each(|(actor_id, seq)| {
-            result.insert(
-                actor_id.clone(),
-                max(*seq, *other.0.get(actor_id).unwrap_or(&(0 as u32))),
-            );
-        });
-        other.0.iter().for_each(|(actor_id, seq)| {
-            result.insert(
-                actor_id.clone(),
-                max(*seq, *self.0.get(actor_id).unwrap_or(&(0 as u32))),
-            );
+    pub fn merge(&mut self, other: &Clock) {
+        other.into_iter().for_each(|(actor_id, seq)| {
+            self.set(actor_id, max(*seq, self.get(actor_id)));
         });
-        Clock(result)
     }
 
-    pub fn is_before_or_concurrent_with(&self, other: &Clock) -> bool {
-        other
-            .0
-            .iter()
-            .all(|(actor_id, seq)| self.0.get(actor_id).unwrap_or(&0) >= seq)
+    pub fn union(&self, other: &Clock) -> Clock {
+        let mut result = self.clone();
+        result.merge(other);
+        result
+    }
+
+    pub fn set(&mut self, actor_id: &ActorID, seq: u32) {
+        if seq == 0 {
+            self.0.remove(actor_id);
+        } else {
+            self.0.insert(actor_id.clone(), seq);
+        }
     }
 
-    pub fn at(&self, actor_id: &ActorID) -> u32 {
+    pub fn get(&self, actor_id: &ActorID) -> u32 {
         *self.0.get(actor_id).unwrap_or(&0)
     }
 
-    /// Returns true if all components of `clock1` are less than or equal to those
-    /// of `clock2` (both clocks given as Immutable.js Map objects). Returns false
-    /// if there is at least one component in which `clock1` is greater than
-    /// `clock2` (that is, either `clock1` is overall greater than `clock2`, or the
-    /// clocks are incomparable).
-    ///
-    /// TODO This feels like it should be a PartialOrd implementation but I
-    /// can't figure out quite what that should look like
-    ///
+
     pub fn less_or_equal(&self, other: &Clock) -> bool {
-        self.0.iter().chain(other.0.iter()).all(|(actor_id, _)| {
-            self.0.get(actor_id).unwrap_or(&0) <= other.0.get(actor_id).unwrap_or(&0)
-        })
+        self.into_iter()
```

Comment:
I would like to check this with Martin, the implementation I wrote, which looks up all actors in both clocks, is taken directly from the javascript implementation. I don't _think_ it's necessary according to the documentation (which was also copied from the javascript implementation), but it's probably worth checking.

Also, can we leave the documentation comment in?

Reply:
Yeah i saw that.  Problem is all the keys in the other clock that are not in this clock are `> 0` and which means we already know they will pass.  Including those keys doesn't seem to change the result for any given two clocks.

Reply:
And all the tests still passed which gave me confidence.

Reply:
Cool, let's merge and just check with Martin when we go through the code this afternoon.